### PR TITLE
feat(api): basic support for openid-connect authentication provider

### PR DESCRIPTION
**Pull request by phsym**
1. Description
This enables basic integration with openid-connect authentication providers like Keycloak or Hydra

2. About tests
Not sure yet how to automate the tests here, but for manual testing :
* Setup Keycloak ([more here](https://www.keycloak.org/getting-started/getting-started-docker))
  * Start a keycloak server in Docker with `docker run -p 4040:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin quay.io/keycloak/keycloak:11.0.1`
  > Map internal port 8080 to 4040 to avoid conflict with CDS UI, in case you run everything locally
  * Login to keycoak web UI as admin
  * Create a realm named "cds"
  * Create a user in this realm, add a password to it
  * Create an openid-connect client named "cds_client", set the CDS urls, and change it from public to confidential. Save, and copy the client secrets
  * Edit cds API config to add open id settings as following
  ```toml
  [api.auth.oidc]
        clientId = "cds_client"
        clientSecret = "<CDS CLIENT SECRET>"
        enabled = true
        signupDisabled = false
        url = "http://<KEYCLOAK ADDRESS>:4040/auth/realms/cds"
  ```
  > `<KEYCLOAK ADDRESS>` can be localhost if you run everything locally
  * Login on CDS UI using the openid-connect button. You'll e redirected to keycloak login screen. After succesfull login, you'll be redirected to CDS screen.

3. Mentions

@ovh/cds

Signed-off-by: phsym <pierre-henri.symoneaux@nokia.com>

### CODE REVIEW

File: engine/api/authentication/oidc/openid.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		return info, sdk.WrapError(err, "Failed to verify ID Token")
+	}
+	tokenClaim := make(map[string]interface{})
+	if err := idToken.Claims(&tokenClaim); err != nil {
+		return info, sdk.WrapError(err, "Cannot unmarshal OIDC claim")
+	}
+
+	if info.ExternalID, ok = tokenClaim["sub"].(string); !ok {
+		return info, sdk.WrapError(sdk.ErrInvalidUser, "Missing OIDC user ID in token claim")
```
Comment (phsym) [inline engine/api/authentication/oidc/openid.go:118]: Please double check the returned error here. Is `sdk.ErrInvalidUser` appropriate ?

**Comment by richardlt** [comment]
Hi, thanks for the PR can you provide some tips about how you test it ?

File: engine/api/authentication/oidc/openid.go (hunk 1 above)
Comment (richardlt) [reply engine/api/authentication/oidc/openid.go:121]: Should be ok, the will return a bad request.

**Comment by phsym** [comment]
Hi, sure ! I just update the PR description with the tips

**Comment by yesnault** [comment]
@phsym Thank you!

Can you take this commit https://github.com/ovh/cds/commit/b9eca49e5191d357d2f35d66db5625d8271a4132 to add openid icon please?

Capture d’écran 2020-08-25 à 19 38 01
Capture d’écran 2020-08-25 à 19 41 59

**Comment by phsym** [comment]
@yesnault Done by merging your branch into mine, but it messed up the DCO check for some reason

**Comment by phsym** [comment]
Oops sorry, I missclicked on close

**Comment by phsym** [comment]
I just added a check on the "email_verified" OIDC claim

File: engine/api/authentication/oidc/openid.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	var info sdk.AuthDriverUserInfo
+
+	ctx2 := context.WithValue(context.Background(), oauth2.HTTPClient, http.DefaultClient)
+	oauth2Token, err := d.OAuth2Config.Exchange(ctx2, req["code"])
+	if err != nil {
+		return info, sdk.WrapError(err, "Failed to exchange token")
+	}
+	rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+	if !ok {
+		return info, sdk.WrapError(err, "No id_token field in oauth2 token")
```
Comment (fsamin) [inline engine/api/authentication/oidc/openid.go:109]: you cant' use `err`. You have to do `sdk.WithStack(fmt.Errorf("no id_token field in oauth2 token"))`.

File: engine/api/authentication/oidc/openid.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	tokenClaim := make(map[string]interface{})
+	if err := idToken.Claims(&tokenClaim); err != nil {
+		return info, sdk.WrapError(err, "Cannot unmarshal OIDC claim")
+	}
+
+	// Check if email is verified.
+	// See standard claims at https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+	if verified, ok := tokenClaim["email_verified"].(bool); !ok || !verified {
+		return info, sdk.WrapError(sdk.ErrInvalidUser, "OIDC user's email not verified")
```
Comment (fsamin) [inline engine/api/authentication/oidc/openid.go:123]: Prefer `sdk.NewErrorFrom(sdk.ErrInvalidUser, "OIDC user's email not verified")` to give the end user more feedback.

File: engine/api/authentication/oidc/openid.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	if info.ExternalID, ok = tokenClaim["sub"].(string); !ok {
+		return info, sdk.WrapError(sdk.ErrInvalidUser, "Missing OIDC user ID in token claim")
```
Comment (fsamin) [inline engine/api/authentication/oidc/openid.go:126]: This kind of error should not raise `ErrInvalidUser ` that means `bad request`. I think that if this piece of data is missing it is a technical issue and should be raised as :
`sdk.WithStack(errors.New("missing OIDC user ID in token claim"))`

File: engine/api/authentication/oidc/openid.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	if info.Username, ok = tokenClaim["preferred_username"].(string); !ok {
+		return info, sdk.WrapError(sdk.ErrInvalidUser, "Missing username in OIDC token claim")
```
Comment (fsamin) [inline engine/api/authentication/oidc/openid.go:130]: same comment as above

File: engine/api/authentication/oidc/openid.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	info.Fullname, _ = tokenClaim["name"].(string)
+	if info.Email, ok = tokenClaim["email"].(string); !ok {
+		return info, sdk.WrapError(sdk.ErrInvalidUser, "Missing user's email in OIDC token claim")
```
Comment (fsamin) [inline engine/api/authentication/oidc/openid.go:135]: same comment as above

**Review by fsamin — CHANGES_REQUESTED** [review]
Thank you very much to your great contribution. Here are some change requests.

File: engine/api/authentication/oidc/openid.go (hunk 2 above)
Comment (phsym) [reply engine/api/authentication/oidc/openid.go:109]: Copy paste mistake, thanks for spotting that

**Comment by phsym** [comment]
@fsamin Corrections provided based on your comments

**Review by yesnault — APPROVED** [review]

File: engine/api/authentication/oidc/openid.go [hunk 7]
```diff
@@ -120,19 +121,19 @@ func (d authDriver) GetUserInfo(ctx context.Context, req sdk.AuthConsumerSigninR
...
+		return info, sdk.NewErrorFrom(sdk.ErrInvalidUser, "OIDC user's email not verified")
 	}
 	if info.ExternalID, ok = tokenClaim["sub"].(string); !ok {
-		return info, sdk.WrapError(sdk.ErrInvalidUser, "Missing OIDC user ID in token claim")
+		return info, sdk.WithStack(errors.New("missing OIDC user ID in token claim"))
 	}
 
 	if info.Username, ok = tokenClaim["preferred_username"].(string); !ok {
-		return info, sdk.WrapError(sdk.ErrInvalidUser, "Missing username in OIDC token claim")
+		return info, sdk.WithStack(errors.New("Missing username in OIDC token claim"))
```
Comment (fsamin) [inline engine/api/authentication/oidc/openid.go:131]:
```suggestion
		return info, sdk.WithStack(errors.New("missing username in OIDC token claim"))
```

File: engine/api/authentication/oidc/openid.go [hunk 8]
```diff
@@ -120,19 +121,19 @@ func (d authDriver) GetUserInfo(ctx context.Context, req sdk.AuthConsumerSigninR
...
 	}
 
 	info.Fullname, _ = tokenClaim["name"].(string)
 	if info.Email, ok = tokenClaim["email"].(string); !ok {
-		return info, sdk.WrapError(sdk.ErrInvalidUser, "Missing user's email in OIDC token claim")
+		return info, sdk.WithStack(errors.New("Missing user's email in OIDC token claim"))
```
Comment (fsamin) [inline engine/api/authentication/oidc/openid.go:136]:
```suggestion
		return info, sdk.WithStack(errors.New("missing user's email in OIDC token claim"))
```

**Review by fsamin — CHANGES_REQUESTED** [review]
I added 2 suggestions following https://github.com/golang/go/wiki/CodeReviewComments#error-strings

**Comment by phsym** [comment]
@fsamin Done

**Review by fsamin — APPROVED** [review]
LGTM

**Review by richardlt — APPROVED** [review]

**Review by sguiheux — APPROVED** [review]

**Pull request merged**